src/interchange/src/avro.rs

```diff
@@ -435,6 +436,94 @@ impl BinlogSchemaIndices {
     }
 }
 
+#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
+pub enum DebeziumDeduplicationStrategy {
+    Ordered,
+    Full,
+}
+
+#[derive(Debug)]
+enum DebeziumDeduplicationState {
+    Ordered {
+        /// Last recorded (pos, row, offset) for each MySQL binlog file.
+        /// Messages that are not ahead of the last recorded pos/row will be skipped.
+        binlog_offsets: HashMap<String, (usize, usize, Option<i64>)>,
+    },
+    Full {
+        seen_offsets: HashMap<String, HashSet<(usize, usize)>>,
+    },
+}
+
+impl DebeziumDeduplicationState {
+    fn new(strat: DebeziumDeduplicationStrategy) -> Self {
+        match strat {
+            DebeziumDeduplicationStrategy::Ordered => DebeziumDeduplicationState::Ordered {
+                binlog_offsets: Default::default(),
+            },
+            DebeziumDeduplicationStrategy::Full => DebeziumDeduplicationState::Full {
+                seen_offsets: Default::default(),
+            },
+        }
+    }
+    #[must_use]
+    fn should_use_record(
+        &mut self,
+        file: &str,
+        pos: usize,
+        row: usize,
+        coord: Option<i64>,
+        debug_name: &str,
+        worker_idx: usize,
+    ) -> bool {
+        match self {
+            DebeziumDeduplicationState::Ordered { binlog_offsets } => {
+                match binlog_offsets.entry(file.to_owned()) {
+                    Entry::Occupied(mut oe) => {
+                        let (old_max_pos, old_max_row, old_offset) = *oe.get();
+                        if old_max_pos > pos || (old_max_pos == pos && old_max_row >= row) {
```

Comment:
For clarity (I'm trying to grok the code) this is the same as `(old_max_pos, old_max_row) >= (pos, row)`?

Reply:
Yep, I'll change it. I think when I wrote this I didn't realize Rust tuples were `Ord`.